### PR TITLE
dune-configurator: respect $PKG_CONFIG

### PR DESCRIPTION
**Pull request by anmonteiro**
- `PKG_CONFIG` is commonly used: there's prior art in [Nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/pkg-config-wrapper/setup-hook.sh#L26), [Cmake](https://cmake.org/cmake/help/latest/module/FindPkgConfig.html)
  - in cross-compilation scenarios, the target environment can set `PKG_CONFIG` to the prefixed executable, e.g. `aarch64-unknown-linux-musl-pkg-config`

fixes https://github.com/ocaml/dune/issues/2572

### CODE REVIEW

**Comment by anmonteiro** [comment]
This allows to remove a `pkg-config` hack in the nix-overlays, for example: https://github.com/nix-ocaml/nix-overlays/pull/716/commits/34f688d65dd267762c20a923f28eba118cebc2a5

**Comment by rgrinberg** [comment]
Thanks for the PR. Some questions about nix/cmake so I understand this variable correctly.

> export PKG_CONFIG${role_post}=@targetPrefix@@baseBinName@

What's role_post in this line?

From the [cmake docs](https://cmake.org/cmake/help/latest/module/FindPkgConfig.html#variables-affecting-behavior)

Should we support the `PKG_CONFIG_EXECUTABLE` variable as well?

> The PKG_CONFIG environment variable can be used as a hint if PKG_CONFIG_EXECUTABLE has not yet been set.

What does it mean to be used as a hint?

In the current implementation, should we allow command line arguments to bw included in this variable? E.g. is this valid: `PKG_CONFIG=pkg-config --foo`?

Finally, could you make a corresponding change to pkg_config.ml in the rules? We also support pkg_config there and it would be good to be consistent.

**Comment by anmonteiro** [comment]
> What's role_post in this line?

Look like it sets `_FOR_BUILD` and `_FOR_TARGET`. see [this](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/setup-hooks/role.bash). It looks only used internally, while `PKG_CONFIG` is exposed to the shell.




> Should we support the `PKG_CONFIG_EXECUTABLE` variable as well?

From what I've read, cmake will look for `PKG_CONFIG` too. [here](https://fuchsia.googlesource.com/third_party/cmake/+/v3.5.1/Modules/FindPkgConfig.cmake) they look for it first. In those cases `PKG_CONFIG_EXECUTABLE` is set from `PKG_CONFIG`. My suggestion is to support only one variable, since `_EXECUTABLE` seems specific to cmake.





> In the current implementation, should we allow command line arguments to bw included in this variable? E.g. is this valid: `PKG_CONFIG=pkg-config --foo`?

I feel like it might be valid indeed, even if there are issues with it in the [wild](https://gitlab.kitware.com/cmake/cmake/-/issues/22305). I'm not sure what to do here. My bias is to support just the binary and adapt it later based on user feedback.


> Finally, could you make a corresponding change to pkg_config.ml in the rules? We also support pkg_config there and it would be good to be consistent.

yup, will do!

**Comment by anmonteiro** [comment]
Added the same logic to `Pkg_config.ml` in rules.

File: src/dune_rules/pkg_config.ml [hunk 1]
```diff
@@ -1,5 +1,11 @@
 open Import
 
+let pkg_config_binary sctx =
+  let env = Super_context.context_env sctx in
+  match Env.get env "PKG_CONFIG" with
```
Comment (rgrinberg) [inline src/dune_rules/pkg_config.ml:5]: We should probably allow overriding this variable. I recommend fetching the environment from `Super_context.env_node`

Comment (anmonteiro) [reply src/dune_rules/pkg_config.ml:6]: Ah yeah, env_node makes sense. I changed to that.

**Review by rgrinberg — APPROVED** [review]
Don't forget to update the CHANGES

**Pull request merged**